### PR TITLE
Unity Testing Framwork MVP (Very experimental)

### PR DESCRIPTION
**Pull request by ChrisGora**
It works.

There are two types of tests:

Non unity and unity

The non unity ones can test a single class if it doesn't use any MonoBehaviour crap

The unity ones are more like integration tests. The single test provided starts up a server and then emits events and checks that the correct events with correct data have been triggered.

It's not exactly easy to write tests but I recommend reusing them

### CODE REVIEW

**Comment by RileyEv** [comment]
85 files :cry:

**Comment by RileyEv** [comment]
Do we need to NSubstitute dll's to be committed?

 and System.Threading.Tasks.Extensions?

**Comment by ChrisGora** [comment]
Yeah.... Otherwise you have to tell everyone to install it manually

File: ProjectFiles/Assets/Scripts/Game/Entity/Player.cs [hunk 1]
```diff
@@ -57,15 +57,15 @@ public class Player {
         /// The spawn method used by the server. 
         /// </summary>
         /// <param name="spawner">Allows the player to get a new spawn location</param>
-        public void Spawn(SpawnLocations spawner) {
+        public virtual void Spawn(SpawnLocations spawner) {
```
Comment (RileyEv) [inline ProjectFiles/Assets/Scripts/Game/Entity/Player.cs:60]: Do we need the virtual?

Comment (RileyEv) [reply ProjectFiles/Assets/Scripts/Game/Entity/Player.cs:60]: was used a few other times aswell

Comment (ChrisGora) [reply ProjectFiles/Assets/Scripts/Game/Entity/Player.cs:60]: Yes, it means a method can be overrided - it's not by default

Comment (RileyEv) [reply ProjectFiles/Assets/Scripts/Game/Entity/Player.cs:60]: but do we override it? I cant see it anywhere

Comment (ChrisGora) [reply ProjectFiles/Assets/Scripts/Game/Entity/Player.cs:60]: Nsubstitute uses overriding under the hood

Comment (RileyEv) [reply ProjectFiles/Assets/Scripts/Game/Entity/Player.cs:60]: Okay just odd that it doesn't error if you take them away

**Review by RileyEv — APPROVED** [review]

**Comment by RileyEv** [comment]
Put stuff about testing here https://github.com/BigMacintosh/Parking/wiki/Unit-testing

**Pull request merged**